cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/GoalViolationDetector.java

```diff
@@ -208,6 +213,11 @@ public void run() {
         provisionResponse.aggregate(goal.provisionResponse());
       }
       _provisionResponse = provisionResponse;
+      // Right-size the cluster (if needed)
+      boolean isRightsized = _provisioner.rightsize(_provisionResponse.recommendationByRecommender());
+      if (isRightsized) {
+        LOG.info("Actions have been taken on the cluster towards rightsizing.");
+      }
```

Comment:
Maybe we want to skip calling `rightsize()` if the status is `UNDECIDED or `RIGHT_SIZED`? 
```suggestion
      if (_provisionResponse.status() == ProvisionStatus.UNDER_PROVISIONED 
              || _provisionResponse.status() == ProvisionStatus.OVER_PROVISIONED) {
        boolean isRightsized = _provisioner.rightsize(_provisionResponse.recommendationByRecommender());
        if (isRightsized) {
          LOG.info("Actions have been taken on the cluster towards rightsizing.");
        }
      }
```

Reply:
I feel this check might be redundant, because if the status is `UNDECIDED` or `RIGHT_SIZED`, the input to `rightsize()` function will be an empty map, which should be a no-op.

Reply:
Got it. I wasn't thinking in functional programming terms. I agree that every implementation of `rightsize()` should first check for empty map and return `false` if it's empty. 
I feel like half of the future contributors would return `true` for this case. We might need to keep an eye out while merging in PR the future.